crates/winch/src/compiler.rs

```diff
@@ -102,17 +116,46 @@ impl wasmtime_environ::Compiler for Compiler {
         types: &ModuleTypes,
         index: DefinedFuncIndex,
     ) -> Result<Box<dyn Any + Send>, CompileError> {
-        let _ = (translation, types, index);
-        todo!()
+        let index = translation.module.func_index(index);
+        let sig = translation.module.functions[index].signature;
+        let ty = &types[sig];
+        let wasm_ty = wasmparser::FuncType::new(
+            ty.params().iter().copied().map(Into::into),
+            ty.returns().iter().copied().map(Into::into),
+        );
+
+        let env = FuncEnv::new(&translation.module, translation.get_types(), &self.isa);
+        let buffer = self
+            .isa
+            .compile_trampoline(&wasm_ty, index.as_u32(), &env, TrampolineKind::NativeToWasm)
+            .map_err(|e| CompileError::Codegen(format!("{:?}", e)))?;
+
+        let compiled_function =
+            CompiledFunction::new(buffer, CompiledFuncEnv {}, self.isa.function_alignment());
+
+        Ok(Box::new(compiled_function))
     }
 
     fn compile_wasm_to_native_trampoline(
         &self,
         translation: &ModuleTranslation<'_>,
         wasm_func_ty: &wasmtime_environ::WasmFuncType,
     ) -> Result<Box<dyn Any + Send>, CompileError> {
-        let _ = (translation, wasm_func_ty);
-        todo!()
+        let wasm_ty = wasmparser::FuncType::new(
+            wasm_func_ty.params().iter().copied().map(Into::into),
+            wasm_func_ty.returns().iter().copied().map(Into::into),
+        );
+
+        let env = FuncEnv::new(&translation.module, translation.get_types(), &self.isa);
+        let buffer = self
+            .isa
+            .compile_trampoline(&wasm_ty, 1u32, &env, TrampolineKind::WasmToNative)
```

Comment:
Instead of manufacturing a dummy `1u32` here could the index of what to call perhaps be a payload of `TrampolineKind`? That way `WasmToNative` would have no payload but the other two would have payloads.